File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        graph=defaultdict(list)
        queue=deque([root])
        while queue:
            curr_node = queue.popleft()
            if curr_node.left:
                graph[curr_node].append(curr_node.left)
                graph[curr_node.left].append(curr_node)
                queue.append(curr_node.left)

            if curr_node.right:
                graph[curr_node].append(curr_node.right)
                graph[curr_node.right].append(curr_node)
                queue.append(curr_node.right)
        res=[]
        visited = set([target])
        # visited.add(target)#use add() for set
        step_queue = deque([(target,0)])
        while step_queue:
            node,step = step_queue.popleft()#debug
            if step==k:
                res.append(node.val)
            else:
                for nodes in graph[node]:#its connected node in a list
                    if nodes not in visited:
                        visited.add(nodes)
                        step_queue.append((nodes,step+1))
        return res

            

            
                
            
        

File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<Integer> distanceK(TreeNode root, int k) {
        List<TreeNode> atCurLevel = new ArrayList<>();
        atCurLevel.add(root);
        for (int i = 0; i < k && 0 < atCurLevel.size(); i++) {
            List<TreeNode> atNextLevel = new ArrayList<>();
            for (int j = 0; j < atCurLevel.size(); j++) {
                TreeNode node = atCurLevel.get(j);

                if (null != node.left) {
                    atNextLevel.add(node.left);
                }

                if (null != node.right) {
                    atNextLevel.add(node.right);
                }
            }
            
            atCurLevel = atNextLevel;
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < atCurLevel.size(); i++) {
            result.add(atCurLevel.get(i).val);
        }

        return result;
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        List<TreeNode> ancestors = new ArrayList<>();
        TreeNode node = root;
        while (target != node) {
            if (null == node.left && null == node.right) {
                TreeNode parent = null;
                while (0 < ancestors.size()) {
                    parent = ancestors.get(ancestors.size() - 1);
                    if (node == parent.left && null != parent.right) {
                        break;
                    }

                    node = parent;
                    ancestors.remove(parent);
                }

                node = parent.right;
                continue;
            }

            ancestors.add(node);
            if (null != node.left) {
                node = node.left;
            } else {
                node = node.right;
            }
        }

        List<Integer> result = distanceK(node, k);
        for (int i = 0; i < k && i < ancestors.size(); i++) {
            TreeNode parent = ancestors.get(ancestors.size() - i - 1);

            if (k - 1 == i) {
                result.add(parent.val);
                continue;
            }

            if (node == parent.left && null != parent.right) {
                result.addAll(distanceK(parent.right, k - i - 2));
            }

            if (node == parent.right && null != parent.left) {
                result.addAll(distanceK(parent.left, k - i - 2));
            }

            node = parent;
        }

        return result;
    }
}